src/Conversion/ONNXToKrnl/Tensor/Unique.cpp

```diff
@@ -0,0 +1,194 @@
+        indices, inverse_indices, counts};
+    rewriter.create<KrnlCallOp>(loc, "omTensorUnique", 8, operands);
+  }
+  return total;
+}
+
+struct ONNXUniqueOpLowering : public ConversionPattern {
+  ONNXUniqueOpLowering(TypeConverter &typeConverter, MLIRContext *ctx)
+      : ConversionPattern(
+            typeConverter, mlir::ONNXUniqueOp::getOperationName(), 1, ctx) {}
+
+  ///
+  /// Intermediate data are presented below for better understanding:
+  ///
+  /// there are 4 subtensors sliced along axis 1 of input_x (shape = (2, 4, 2)):
+  /// A: [[1, 1], [1, 1]],
+  ///    [[0, 1], [0, 1]],
+  ///    [[2, 1], [2, 1]],
+  ///    [[0, 1], [0, 1]].
+  ///
+  /// there are 3 unique subtensors:
+  /// [[1, 1], [1, 1]],
+  /// [[0, 1], [0, 1]],
+  /// [[2, 1], [2, 1]].
+  ///
+  /// sorted unique subtensors:
+  /// B: [[0, 1], [0, 1]],
+  ///    [[1, 1], [1, 1]],
+  ///    [[2, 1], [2, 1]].
+  ///
+  /// output_Y is constructed from B:
+  /// [[[0. 1.], [1. 1.], [2. 1.]],
+  /// [[0. 1.], [1. 1.], [2. 1.]]]
+  ///
+  /// output_indices is to map from B to A:
+  /// [1, 0, 2]
+  ///
+  /// output_inverse_indices is to map from A to B:
+  /// [1, 0, 2, 0]
+  ///
+  /// output_counts = [2 1 1]
+  ///
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    ONNXTopKOpAdaptor operandAdaptor(operands, op->getAttrDictionary());
+    ONNXUniqueOp uniqueOp = llvm::cast<ONNXUniqueOp>(op);
+    Location loc = op->getLoc();
+    MultiDialectBuilder<KrnlBuilder, IndexExprBuilderForKrnl, MathBuilder,
+        MemRefBuilder>
+        create(rewriter, loc);
+    IndexExprScope scope(create.krnl);
+    ONNXUniqueOpShapeHelper shapeHelper(op, operands, &create.krnlIE);
+    Value X = operandAdaptor.getX();
+    ArrayRef<int64_t> xShape = getShape(X.getType());
+    Type elementType = X.getType().cast<MemRefType>().getElementType();
+    int64_t rank = create.krnlIE.getShapedTypeRank(X);
+    int64_t sorted = operandAdaptor.getSorted();
+    std::optional<int64_t> optionalAxis = uniqueOp.getAxis();
+    //
+    // Get axis value as a positive integer if axis attribute is specified.
+    // If set "-1" if the axis attribute is not specified.
+    //
+    int64_t axis = -1;
+    if (optionalAxis.has_value()) {
+      axis = optionalAxis.value();
+      axis = axis < 0 ? axis + rank : axis;
+      assert(axis >= 0 && axis < rank && "axis is out of bound");
+    }
+    //
+    // Emit a Unique call to get the outputs to count unique
+    // slices(subtensors) of X along axis at first.
+    //
+    Type indexTy = rewriter.getIndexType();
+    Value iZero = create.math.constantIndex(0);
+    Value uniqueCount = create.mem.alloca(MemRefType::get({}, indexTy));
+    create.krnl.store(iZero, uniqueCount, {});
+    Value noneValue;
+    emitArgUnique(rewriter, loc, uniqueCount, X, axis, /*sorted=*/sorted,
+        noneValue, noneValue, noneValue, noneValue, /*count_only=*/true);
```

Comment:
I guess the noneValue will be translated into null pointer. Right?

Reply:
Yes. Correct.